src/BuiltInTools/DotNetDeltaApplier/HotReloadAgent.cs

```diff
@@ -45,116 +44,211 @@ private void OnAssemblyLoad(object? _, AssemblyLoadEventArgs eventArgs)
+            {
+                handlerActions.UpdateApplication.Add(CreateAction(updateApplication));
+                methodFound = true;
+            }
+
+            if (!methodFound)
+            {
+                _log($"No invokable methods found on metadata handler type '{handlerType}'. " +
+                    $"Allowed methods are ClearCache, UpdateApplication");
+            }
+
+            Action<Type[]?> CreateAction(MethodInfo update)
+            {
+                Action<Type[]?> action = update.CreateDelegate<Action<Type[]?>>();
+                return types =>
+                {
+                    try
                     {
-                        before.Add(CreateAction(beforeUpdate));
-                        methodFound = true;
+                        action(types);
                     }
-
-                    if (GetUpdateMethod(handlerType, "AfterUpdate") is MethodInfo afterUpdate)
+                    catch (Exception ex)
                     {
-                        after.Add(CreateAction(afterUpdate));
-                        methodFound = true;
+                        _log($"Exception from '{action}': {ex}");
                     }
+                };
+            }
 
-                    if (!methodFound)
+            MethodInfo? GetUpdateMethod(Type handlerType, string name)
+            {
+                if (handlerType.GetMethod(name, BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static, new[] { typeof(Type[]) }) is MethodInfo updateMethod &&
+                    updateMethod.ReturnType == typeof(void))
+                {
+                    return updateMethod;
+                }
+
+                foreach (MethodInfo method in handlerType.GetMethods(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static | BindingFlags.Instance))
+                {
+                    if (method.Name == name)
                     {
-                        _log($"No BeforeUpdate or AfterUpdate method found on '{handlerType}'.");
+                        _log($"Type '{handlerType}' has method '{method}' that does not match the required signature.");
+                        break;
                     }
+                }
+
+                return null;
+            }
+        }
+
+        internal static List<Assembly> TopologicalSort(ReadOnlySpan<Assembly> assemblies)
+        {
+            var sortedAssemblies = new List<Assembly>(assemblies.Length);
 
-                    Action<Type[]?> CreateAction(MethodInfo update)
+            var visited = new HashSet<string>(StringComparer.Ordinal);
+
+            foreach (var assembly in assemblies)
+            {
+                Visit(assemblies, assembly, sortedAssemblies, visited);
+            }
+
+            static void Visit(ReadOnlySpan<Assembly> assemblies, Assembly assembly, List<Assembly> sortedAssemblies, HashSet<string> visited)
+            {
+                var assemblyIdentifier = assembly.FullName ?? assembly.ToString();
```

Comment:
```suggestion
                var assemblyIdentifier = assembly.FullName!;
```
Runtime loaded assemblies always have non-empty full names. This is just nullability annotations not being helpful.